### ISSUE TITLE
iscsi_tgtd hanging, possibly related to hotplug

### ISSUE THREAD

**Issue by yaahallo** [spdk/spdk#725]
We're integrating iscsi_tgtd into our product and some of our existing test cases seem to cause iscsi_tgtd to hang. Below is a pstack of a hung iscsi_tgtd process.

```
[192.168.13.22 (54c5731c) scale 21:40:19]# pstack $(pidof iscsi_tgtd)
Thread 2 (Thread 0x7f88a07b7700 (LWP 6992)):
#0  0x00007f88a0ce3183 in epoll_wait () from /lib64/libc.so.6
#1  0x000000000048d994 in eal_intr_thread_main ()
#2  0x00007f88a0fb8e25 in start_thread () from /lib64/libpthread.so.0
#3  0x00007f88a0ce2bad in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f88a225a9c0 (LWP 6984)):
#0  0x000000000045b2d8 in _spdk_iscsi_conn_abort_queued_datain_task ()
#1  0x000000000045c035 in _spdk_iscsi_op_abort_task_set ()
#2  0x00000000004726cc in spdk_thread_poll ()
#3  0x0000000000447912 in _spdk_reactor_run ()
#4  0x0000000000447dcd in spdk_reactors_start ()
#5  0x0000000000446c4e in spdk_app_start ()
#6  0x0000000000405bfe in main ()
```
The tests that reproduce the issue most consistently are the one that simulate various forms of drive failures, such as faking io errors so the system decides to evacuate and remove a drive, or kicking the drive from the kernel to simulate the drive being pulled physically.

We're using spdk version 19.01

we're using aio as our bdev module

I'm currently working on figuring out exactly how to reproduce the failure. As soon as I end up getting a more clear path to repro than "grind problematic tests until the failure happens" I'll post it.

**Comment by shuhei-matsumoto** [comment]
Hi @yaahallo 

Thanks for reporting this issue.  It will be great if you can share with us how to reproduce the issue.

SPDK test pool have found the issue that hot removal is not completed.
https://github.com/spdk/spdk/issues/715 

Improving LUN reset will not be matured yet because this is added to spdk 19.01. We will have to improve the quality of iSCSI hotplug.

Thanks,
Shuhei

**Comment by jimharris** [comment]
@yaahallo was on IRC and described this a bit.  I think key detail is that her test is deleting the /dev/xyz node via sysfs while aio module has it open.  SPDK aio module doesn't try to handle this case.  It's all related to general reset handling with the aio module - to be honest, there's no good way to really do a reset with the Linux aio interfaces.

If someone wants to take a look at this, I think the following should likely reproduce the problem:

1) Create AIO bdev with some kernel block device - it doesn't really matter which one, but I'd suggest not using a ramdisk so that there's a delay between submission and completion.
2) Create iSCSI target node on this AIO bdev
3) Connect to target node with kernel iSCSI initiator
4) Run high queue depth FIO
5) Delete the kernel block device using sysfs (echo 1 > /sys/block/%s/device/delete)

I think the solution is that on aio bdev reset, the aio module needs to:
1) Delete the AIO context
2) Close the fd
3) Reopen the fd
4) Resubmit the I/O

If we can't reopen the fd (step #3), then we can complete the reset with failure and immediately unregister the bdev.

**Comment by shuhei-matsumoto** [comment]
@jimharris 

Thank you for writing down the issue and possible fix very carefully. I can't promise I will post any update but I'll work on this today in my time zone.

**Comment by yaahallo** [comment]
I succeeded in reproducing the issue with the steps described above, though it didn't occur until my second attempt.

**Comment by shuhei-matsumoto** [comment]
@yaahallo 

I added three aio bdevs for sdd, sde, sdf to an iscsi target.
I ran FIO from another server with 4KB random read and iodepth=1024.
But I removed sdd, sde, and sdf during FIO running, but couldn't reproduce the issue.

I'll try more but can you share the setting of FIO and SPDK iSCSI target?

Config file is the following
```
[global]
  rw=randread
  bs=4k
  numjobs=3
  iodepth=1024
  runtime=120
  group_reporting=1
  name=test
  direct=1
  ioengine=libaio
  loops=1
  invalidate=1
  ramp_time=18
  randrepeat=0
  norandommap
  exitall
  cpus_allowed_policy=split
  thread=1
  lockfile=none

[job1]
  filename=/dev/sdc

[job2]
  filename=/dev/sdd

[job3]
  filename=/dev/sde

```

**Comment by yaahallo** [comment]
@shuhei-matsumoto I'm trying to figure out how to reproduce the issue without any of our system specific code running and am currently blocked by the fact that I hardly know how to use fio and when I try to connect to iscsi_tgtd directly rather than through our custom ioengine fio is just exiting immediately.

The fio command I actually ran when I successfully reproduced the issue manually was.

```/usr/bin/fio --ioengine=<internallibstuff> --thread --gtod_reduce=0 --norandommap --randrepeat=1 --name=drsdDemo --blockalign=4K --size=50G --bs=32k --rw=randwrite --iodepth=8 --filename="$vsd_uuid"```

I'm not sure how useful this is though.

When I try following the steps on https://spdk.io/doc/iscsi.html to connect to our bdevs with iscsiadm and then run the same fio command above with the filename changed to `/dev/sde` and ioengine changed to libaio the command exits immediately with status code 0.

as for spdk config, here is our json config 

```
{
  "subsystems": [
    {
      "subsystem": "copy",
      "config": null
    },
    {
      "subsystem": "interface",
      "config": null
    },
    {
      "subsystem": "net_framework",
      "config": null
    },
    {
      "subsystem": "bdev",
      "config": [
        {
          "params": {
            "bdev_io_pool_size": 65536,
            "bdev_io_cache_size": 256
          },
          "method": "set_bdev_options"
        },
        {
          "params": {
            "retry_count": 4,
            "timeout_us": 0,
            "nvme_adminq_poll_period_us": 1000000,
            "action_on_timeout": "none"
          },
          "method": "set_bdev_nvme_options"
        },
        {
          "params": {
            "enable": false,
            "period_us": 100000
          },
          "method": "set_bdev_nvme_hotplug"
        },
        {
          "params": {
            "block_size": 512,
            "name": "Slot_1",
            "filename": "/dev/scale/slot1data"
          },
          "method": "construct_aio_bdev"
        },
        {
          "params": {
            "block_size": 512,
            "name": "Slot_2",
            "filename": "/dev/scale/slot2data"
          },
          "method": "construct_aio_bdev"
        },
        {
          "params": {
            "block_size": 512,
            "name": "Slot_3",
            "filename": "/dev/scale/slot3data"
          },
          "method": "construct_aio_bdev"
        },
        {
          "params": {
            "block_size": 512,
            "name": "Slot_0",
            "filename": "/dev/scale/slot0data"
          },
          "method": "construct_aio_bdev"
        }
      ]
    },
    {
      "subsystem": "scsi",
      "config": null
    },
    {
      "subsystem": "nbd",
      "config": []
    },
    {
      "subsystem": "iscsi",
      "config": [
        {
          "params": {
            "allow_duplicated_isid": false,
            "default_time2retain": 60,
            "mutual_chap": false,
            "require_chap": false,
            "immediate_data": true,
            "node_base": "iqn.2018-02.scale",
            "nop_in_interval": 30,
            "max_connections_per_session": 2,
            "first_burst_length": 8192,
            "max_queue_depth": 64,
            "nop_timeout": 30,
            "chap_group": 0,
            "max_sessions": 128,
            "error_recovery_level": 0,
            "disable_chap": false,
            "auth_file": "/usr/local/etc/spdk/auth.conf",
            "min_connections_per_core": 4,
            "default_time2wait": 2
          },
          "method": "set_iscsi_options"
        },
        {
          "params": {
            "portals": [
              {
                "cpumask": "0x1",
                "host": "192.168.3.12",
                "port": "3260"
              }
            ],
            "tag": 1
          },
          "method": "add_portal_group"
        },
        {
          "params": {
            "initiators": [
              "ANY"
            ],
            "tag": 1,
            "netmasks": [
              "192.168.3.12/24"
            ]
          },
          "method": "add_initiator_group"
        },
        {
          "params": {
            "luns": [
              {
                "lun_id": 0,
                "bdev_name": "Slot_1"
              }
            ],
            "mutual_chap": false,
            "name": "iqn.2018-02.scale:192-168-3-12-Slot-1",
            "alias_name": "192.168.3.12_Slot_1",
            "require_chap": false,
            "chap_group": 0,
            "pg_ig_maps": [
              {
                "ig_tag": 1,
                "pg_tag": 1
              }
            ],
            "data_digest": false,
            "disable_chap": false,
            "header_digest": false,
            "queue_depth": 64
          },
          "method": "construct_target_node"
        },
        {
          "params": {
            "luns": [
              {
                "lun_id": 0,
                "bdev_name": "Slot_2"
              }
            ],
            "mutual_chap": false,
            "name": "iqn.2018-02.scale:192-168-3-12-Slot-2",
            "alias_name": "192.168.3.12_Slot_2",
            "require_chap": false,
            "chap_group": 0,
            "pg_ig_maps": [
              {
                "ig_tag": 1,
                "pg_tag": 1
              }
            ],
            "data_digest": false,
            "disable_chap": false,
            "header_digest": false,
            "queue_depth": 64
          },
          "method": "construct_target_node"
        },
        {
          "params": {
            "luns": [
              {
                "lun_id": 0,
                "bdev_name": "Slot_3"
              }
            ],
            "mutual_chap": false,
            "name": "iqn.2018-02.scale:192-168-3-12-Slot-3",
            "alias_name": "192.168.3.12_Slot_3",
            "require_chap": false,
            "chap_group": 0,
            "pg_ig_maps": [
              {
                "ig_tag": 1,
                "pg_tag": 1
              }
            ],
            "data_digest": false,
            "disable_chap": false,
            "header_digest": false,
            "queue_depth": 64
          },
          "method": "construct_target_node"
        },
        {
          "params": {
            "luns": [
              {
                "lun_id": 0,
                "bdev_name": "Slot_0"
              }
            ],
            "mutual_chap": false,
            "name": "iqn.2018-02.scale:192-168-3-12-Slot-0",
            "alias_name": "192.168.3.12_Slot_0",
            "require_chap": false,
            "chap_group": 0,
            "pg_ig_maps": [
              {
                "ig_tag": 1,
                "pg_tag": 1
              }
            ],
            "data_digest": false,
            "disable_chap": false,
            "header_digest": false,
            "queue_depth": 64
          },
          "method": "construct_target_node"
        }
      ]
    }
  ]
}
```

**Comment by jimharris** [comment]
Hi @shuhei-matsumoto 

Just to confirm - the I/O path should be:

FIO -> kernel iSCSI block device -> SPDK iSCSI target -> SPDK AIO bdev -> kernel ??? block device

Then while this is running, delete "kernel ??? block device".

Just wanted to confirm that you are not deleting the "kernel iSCSI block device".

**Comment by jimharris** [comment]
@shuhei-matsumoto reports on IRC that he is using an HDD as "kernel ??? block device" and removing it via sysfs as suggested above.

@yaahallo - what type of device are you attaching to the AIO bdev?  This could be the configuration difference.

**Comment by yaahallo** [comment]
They're all HDD sata drives

**Comment by shuhei-matsumoto** [comment]
@yaahallo @jimharris

Umm, I have tried to reproduce the issue but my trial has failed so far.
FIO stopped due to IO error as expected but the iSCSI target continues to respond to RPCs even after removal of device.

**Comment by shuhei-matsumoto** [comment]
@yaahallo  @jimharris 

Previous experience started from
 https://review.gerrithub.io/#/c/spdk/spdk/+/386837/
 https://lists.01.org/pipermail/spdk/2017-November/000995.html
was reminded to me.

When we don't specify O_DIRECT + O_DSYNC but specify only O_DIRECT, io_getevents continues to return success even after hot removal. I checked both system tests and code inspection.

Will you try add O_DSYNC if you can?  I will try too when I reproduce.

```
diff --git a/lib/bdev/aio/bdev_aio.c b/lib/bdev/aio/bdev_aio.c
index a6d4a0010..320b50d03 100644
--- a/lib/bdev/aio/bdev_aio.c
+++ b/lib/bdev/aio/bdev_aio.c
@@ -123,7 +123,7 @@ bdev_aio_open(struct file_disk *disk)
 {
        int fd;

-       fd = open(disk->filename, O_RDWR | O_DIRECT);
+       fd = open(disk->filename, O_RDWR | O_DIRECT|O_DSYNC);
        if (fd < 0) {
                /* Try without O_DIRECT for non-disk files */
                fd = open(disk->filename, O_RDWR);

```

Besides, Jim's idea looks good but I need to understand more to implement it.

**Comment by yaahallo** [comment]
The hang is still reproducible with the change mentioned above. :/

I've sofar been unable to reproduce the failure without our system specific code running, or even a subset of it disabled. I did manage to get a segmentation_fault so i figured I'd show you where I'm at right now incase this gives you guys any ideas.


```bash
#! /bin/bash
# abuse_iscsi_tgtd.sh

set -ex

function rescan_drives() {
    while IFS= read -r -d '' dir; do
        /bin/echo - - - >"$dir/scan"
    done < <(find /sys/class/scsi_host -name "host*" -print0)
}

service iscsi_tgtd restart
rescan_drives
sleep 2
rpc.py add_portal_group 1 "$BACKPLANE_ENV":3260
rpc.py add_initiator_group 1 ANY "$BACKPLANE_ENV"/24
rpc.py construct_aio_bdev /dev/sda3 Slot_0 512
rpc.py construct_aio_bdev /dev/scale/slot1data Slot_1 512
rpc.py construct_aio_bdev /dev/scale/slot2data Slot_2 512
rpc.py construct_aio_bdev /dev/scale/slot3data Slot_3 512
rpc.py construct_target_node "$BACKPLANE_ENV"-Slot-0 "$BACKPLANE_ENV"_Slot_0 Slot_0:0 1:1 128
rpc.py construct_target_node "$BACKPLANE_ENV"-Slot-1 "$BACKPLANE_ENV"_Slot_1 Slot_1:0 1:1 128
rpc.py construct_target_node "$BACKPLANE_ENV"-Slot-2 "$BACKPLANE_ENV"_Slot_2 Slot_2:0 1:1 128
rpc.py construct_target_node "$BACKPLANE_ENV"-Slot-3 "$BACKPLANE_ENV"_Slot_3 Slot_3:0 1:1 128

while true; do
    echo 1 >/sys/block/sda/device/delete
    sleep 1
    ls /dev/ | grep -E "^sd[a-z]$"
    sleep 1
    rpc.py delete_target_node iqn.2018-02.scale:"$BACKPLANE_ENV"-Slot-0
    sleep 1
    rpc.py delete_aio_bdev Slot_0
    sleep 10
    rescan_drives
    ls /dev/sd?

    sleep 1
    rpc.py construct_aio_bdev /dev/sda3 Slot_0 512
    sleep 1
    rpc.py construct_target_node "$BACKPLANE_ENV"-Slot-0 "$BACKPLANE_ENV"_Slot_0 Slot_0:0 1:1 128
    sleep 2
done
```

This is my current attempt at reproducing what our system is doing when the hang manifests. I have an fio job running in the background through our internal io stack. I'm fairly certain though that without our management daemon going and recovering the drives and notifying the io stack that the drives are ready for io no io is actually going from the fio to the drives on the node where im fiddling with iscsi_tgtd.

I'd like to figure out how to cut out our io stack entirely but for some reason the behavior of the drive letter assignment differs depending on if i do fio directly against a dev file created by iscsiadm vs if I do it through our internal io stack. When im doing fio through scribed it will always readd the drive with the same letter it originally had (/dev/sda), but when I do fio directly to the drive through the kernel initiator the drive will come back on a new unused letter (/dev/sde).

I think figuring out why this behavior is affected by what ioengine I use with fio is critical to moving forward towards you guys reproducing the hang.

And incase its useful, heres the backtrace I got from the segfault I mentioned in the beginning

```(gdb) bt
#0  0x0000000000453cc8 in spdk_bdev_readv_blocks ()
#1  0x0000000000453e2a in spdk_bdev_readv ()
#2  0x000000000046e507 in spdk_bdev_scsi_readwrite ()
#3  0x000000000046ec35 in spdk_bdev_scsi_process_block ()
#4  0x000000000046ef4d in spdk_bdev_scsi_execute ()
#5  0x000000000046d3a4 in spdk_scsi_lun_execute_tasks ()
#6  0x000000000045c9e4 in spdk_iscsi_conn_handle_queued_datain_tasks ()
#7  0x0000000000460a34 in spdk_iscsi_execute ()
#8  0x00000000004584b2 in spdk_iscsi_conn_sock_cb ()
#9  0x00000000004795e4 in spdk_sock_group_poll_count ()
#10 0x00000000004677e4 in spdk_iscsi_poll_group_poll ()
#11 0x0000000000472657 in spdk_thread_poll ()
#12 0x0000000000447912 in _spdk_reactor_run ()
#13 0x0000000000447dcd in spdk_reactors_start ()
#14 0x0000000000446c4e in spdk_app_start ()
#15 0x0000000000405bfe in main ()
```

**Comment by shuhei-matsumoto** [comment]
@yaahallo 

Do you still observe the issue when calling the following order?
-  delete /dev/sda
-  delete_aio_bdev
-  delete_target_node

**Comment by yaahallo** [comment]
@shuhei-matsumoto 

I changed the order but still managed to reproduce the issue.

However, in verifying that I was running what i thought I was I found that the node that experienced the failure never executed either delete commands. I'm digging more into the timing of the commands that we did run. For the most part we just execute a repeating pattern of get_bdevs, get_portal_groups,  get_intiator_groups, and get_target_nodes. And we did see some construct_aio_bdev, construct_target_node calls early in the test which surprises me so I'm going to look into it, but this is good, knowing that its not related to those commands changes how I'll approach finding the repro for you guys.

**Comment by shuhei-matsumoto** [comment]
@yaahallo 

Thank you providing detailed scripts and analysis.
You say delete commands are unrelated but 
  echo 1 >/sys/block/sda/device/delete
is still the key trigger to cause the segmentation fault?

**Comment by yaahallo** [comment]
@shuhei-matsumoto 

The main issue isnt the segfault, that was just found in my attempt to reproduce the hang. The main issue is spdk getting hung and needing to be killed.

The specific test that I  ran doesn't use the `echo 1 >/sys/block/sda/device/delete` directly though in the past I've reproduced the issue running said command manually. I'll setup a simple test that just kicks the drive with the command you mentioned and double check / verify that it has the same behavior and behaves as I've described and likewise does not call either of the rpc.py delete commands.

**Comment by shuhei-matsumoto** [comment]
Thread 1 (Thread 0x7f88a225a9c0 (LWP 6984)):
#0  0x000000000045b2d8 in _spdk_iscsi_conn_abort_queued_datain_task ()
#1  0x000000000045c035 in _spdk_iscsi_op_abort_task_set ()

iscsi task abort will not complete if there is any pending task. However I haven't found any code that falls in infinite loop when we remove block device. Adding any trace to related code may be helpful and I'm investigating it now.

**Comment by darsto** [comment]
@shuhei-matsumoto you might be interested - our iscsi+fio hotremove test also occasionally hangs on CI. See here https://dqtibwqq6s6ux.cloudfront.net/results/autotest-per-patch/builds/27755/archive/RBD_autotest/build.log

Could this be the same issue?

**Comment by shuhei-matsumoto** [comment]
@darsto maybe, and I'm interested in that. I've started to investigate that but the priority of that is not so high yet, though.

**Comment by shuhei-matsumoto** [comment]
next steps:
-  push patches to dump the state when the issue occurs.
-  patches may be usable for #715 too. 
-  Jane will cherry-pick the patches, collect data in her environment, and share with us.

-  Jane have used libiscsi as initiator, so running FIO or bdevperf on top of libiscsi bdev and connecting libiscsi bdev to SPDK iSCSI target may be helpful to reproduce the issue. We will work on this next if the first doesn't work.

**Comment by shuhei-matsumoto** [comment]
@yaahallo 

I posted three patches to collect data when the issue occurs
They are expected to output states information once per second after 20 seconds timeout.
https://review.gerrithub.io/#/c/spdk/spdk/+/450133/
https://review.gerrithub.io/#/c/spdk/spdk/+/450134/
https://review.gerrithub.io/#/c/spdk/spdk/+/450135/

Will you apply the patches?

**Comment by yaahallo** [comment]
I've applied those three patches. I'm sad to report none of the new traces seemed to be triggered. Here are the logs I did receive during the most recent repro.

[iscsi_tgtd.log.gz](https://github.com/spdk/spdk/files/3049072/iscsi_tgtd.log.gz)

I'm digging  into what was done and whats happening in the hung iscsi target to see if I can't figure out what went wrong.

**Comment by yaahallo** [comment]
I kinda hackily traced what its doing using this script

```bash
#!/usr/bin/env bash
set -euo pipefail
IFS=$'\n\t'

pstack "$(pidof iscsi_tgtd)" >/tmp/last.stack
while true; do
    pstack "$(pidof iscsi_tgtd)" >/tmp/curr.stack
    diff /tmp/last.stack /tmp/curr.stack | tail -n1 >>/tmp/changes.log || true
    mv -f /tmp/curr.stack /tmp/last.stack
done
```
and parsed it with this `cat /tmp/changes.log | awk '{print $NF}' | sort -u`

End Result

```
iscsi.c:3180
iscsi.c:3181
iscsi.c:3184
iscsi.c:3192
```

which is summarized as the following on my current source copy


```
_spdk_iscsi_conn_abort_queued_datain_task(struct spdk_iscsi_conn *conn,
...
	while (conn->data_in_cnt < MAX_LARGE_DATAIN_PER_CONNECTION) {
		assert(task->current_datain_offset <= task->scsi.transfer_len);

		/* If no IO is submitted yet, just abort the primary task. */
		if (task->current_datain_offset == 0) {
...
		if (task->current_datain_offset < task->scsi.transfer_len) {
...
}```

**Comment by yaahallo** [comment]
from gdb

```
>>> p task->current_datain_offset
$2 = 1048576
>>> p task->scsi.transfer_len
$3 = 1048576
```

**Comment by jimharris** [comment]
After seeing this last bit of data from @yaahallo, it was very clear where the problem was.  The solution was fairly simple so I went ahead and made a patch.

https://review.gerrithub.io/c/spdk/spdk/+/450301

@yaahallo reports this fixes the issue on her end.

**Comment by yaahallo** [comment]
Fixes the hang, it's hold off on closing this until we have done more testing with the fix applied. We're also going to log when we go through the that third code path since you said you're unsure how we got into that state. Hopefully we can at least know when we avoided the issu and maybe correlate that with other positive or negative behavior.

**Comment by jimharris** [comment]
It's still not clear to me how we get into this hang condition.  @shuhei-matsumoto - maybe you have some insight on this.  The data from @yaahallo shows that current_datain_offset and transfer_len both equal 1MB (1048576).  This is the max I/O size in the workload according to the fio script that Jane shared over IRC.

One of two things must have happened:
1) _iscsi_conn_abort_queued_datain_task() was called and the task already has current_datain_offset == transfer_len.  But I don't see how that can happen - spdk_iscsi_conn_handle_queued_datain_tasks() should have removed the task from the TAILQ if this happened, which means it could not get passed to this abort function.
2) _iscsi_conn_abort_queued_datain_task() increments the current_datain_offset to match transfer_len.  But there's a check for whether these are equal, and will remove it from the TAILQ and return immediately if they are equal.

**Comment by shuhei-matsumoto** [comment]
@yaahallo @jimharris 

I should have the logic of _iscsi_conn_abort_queued_datain_task() match spdk_iscsi_conn_handle_queued_datain_tasks(), but couldn't it because I could not find the path which needed the fix by Jim.

Simple UT code could not find any error for the above case #2.

```
diff --git a/test/unit/lib/iscsi/iscsi.c/iscsi_ut.c b/test/unit/lib/iscsi/iscsi.c/iscsi_ut.c
index 8d9a09e84..63d2a3117 100644
--- a/test/unit/lib/iscsi/iscsi.c/iscsi_ut.c
+++ b/test/unit/lib/iscsi/iscsi.c/iscsi_ut.c
@@ -1026,6 +1026,24 @@ abort_queued_datain_task_test(void)

        spdk_iscsi_task_cpl(&task->scsi);

+       /* Abort a large read I/O */
+       conn.data_in_cnt = 1;
+
+       task = spdk_iscsi_task_get(&conn, NULL, NULL);
+       SPDK_CU_ASSERT_FATAL(task != NULL);
+       TAILQ_INSERT_TAIL(&conn.queued_datain_tasks, task, link);
+
+       task->scsi.transfer_len = SPDK_BDEV_LARGE_BUF_MAX_SIZE * 16;
+       task->current_datain_offset = SPDK_BDEV_LARGE_BUF_MAX_SIZE;
+
+       rc = _iscsi_conn_abort_queued_datain_task(&conn, task);
+       CU_ASSERT(rc == 0);
+       CU_ASSERT(TAILQ_EMPTY(&conn.queued_datain_tasks));
+       CU_ASSERT(task->current_datain_offset == SPDK_BDEV_LARGE_BUF_MAX_SIZE * 16);
+       CU_ASSERT(conn.data_in_cnt == 16);
+
+       spdk_iscsi_task_cpl(&task->scsi);
+
        /* Queue three data in tasks and abort each task sequentially */
        task = spdk_iscsi_task_get(&conn, NULL, NULL);
        SPDK_CU_ASSERT_FATAL(task != NULL);
```

**Comment by shuhei-matsumoto** [comment]
@yaahallo @jimharris 

Will you apply the following change on top of the Jim's fix, and then enable assert by  ./configure --enable-debug and make, and reproduce the issue by using the debug built binary?

If there is the case #1, SPDK iSCSI target will abort.

**Comment by shuhei-matsumoto** [comment]
@yaahallo

I missed to attach the change.  Will you try the following?
1. apply the following change.
2. ./configure --enable-debug
3. make
4. use the binary and reproduce the issue.

```
diff --git a/lib/iscsi/iscsi.c b/lib/iscsi/iscsi.c
index 13779b36b..d93a53bad 100644
--- a/lib/iscsi/iscsi.c
+++ b/lib/iscsi/iscsi.c
@@ -2872,7 +2872,7 @@ int spdk_iscsi_conn_handle_queued_datain_tasks(struct spdk_iscsi_conn *conn)
        while (!TAILQ_EMPTY(&conn->queued_datain_tasks) &&
               conn->data_in_cnt < MAX_LARGE_DATAIN_PER_CONNECTION) {
                task = TAILQ_FIRST(&conn->queued_datain_tasks);
-               assert(task->current_datain_offset <= task->scsi.transfer_len);
+               assert(task->current_datain_offset < task->scsi.transfer_len);

                if (task->current_datain_offset == 0) {
                        task->scsi.lun = spdk_scsi_dev_get_lun(conn->dev, task->lun_id);
@@ -3291,7 +3291,7 @@ _iscsi_conn_abort_queued_datain_task(struct spdk_iscsi_conn *conn,
        uint32_t remaining_size;

        while (conn->data_in_cnt < MAX_LARGE_DATAIN_PER_CONNECTION) {
-               assert(task->current_datain_offset <= task->scsi.transfer_len);
+               assert(task->current_datain_offset < task->scsi.transfer_len);

                /* If no IO is submitted yet, just abort the primary task. */
                if (task->current_datain_offset == 0) {

```

**Comment by yaahallo** [comment]
I am applying the above change now and attempting the repro while watching for an abort rather than a hang.

**Comment by shuhei-matsumoto** [comment]
Thank you, and if debug build prevents the issue from occurring, will you try to change the assert to something like if (task->current_datain_offset < task->scsi.transfer_len) { SPDK_ERRLOG("Unexpectedly current_datain_offset is equal to scsi_transfer_len); }

**Comment by yaahallo** [comment]
The issue reproduced rather easily

```─── Stack ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
[4] from 0x0000000000488802 in spdk_iscsi_conn_handle_queued_datain_tasks+84 at iscsi.c:2769
arg conn = 0x7ff72f021e00
[+]
─── Expressions ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
─── Source ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Cannot display "/local/scale-product/thirdparty/spdk/spdk/lib/iscsi/iscsi.c" ([Errno 2] No such file or directory: '/local/scale-product/thirdparty/spdk/spdk/lib/iscsi/iscsi.c')
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
>>> bt
#0  0x00007ff738621277 in raise () from /lib64/libc.so.6
#1  0x00007ff738622968 in abort () from /lib64/libc.so.6
#2  0x00007ff73861a096 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007ff73861a142 in __assert_fail () from /lib64/libc.so.6
#4  0x0000000000488802 in spdk_iscsi_conn_handle_queued_datain_tasks (conn=0x7ff72f021e00) at iscsi.c:2769
#5  0x000000000047ec41 in process_read_task_completion (conn=0x7ff72f021e00, task=0x20002a73d4c0, primary=0x20002a6b5bc0) at conn.c:1007
#6  0x000000000047ece6 in spdk_iscsi_task_cpl (scsi_task=0x20002a73d4c0) at conn.c:1024
#7  0x000000000049df6a in spdk_scsi_lun_complete_task (lun=0x1f266d0, task=0x20002a73d4c0) at lun.c:50
#8  0x00000000004a1838 in spdk_bdev_scsi_task_complete_cmd (bdev_io=0x2000075706c0, success=true, cb_arg=0x20002a73d4c0) at scsi_bdev.c:1282
#9  0x0000000000478698 in _spdk_bdev_io_complete (ctx=0x2000075706c0) at bdev.c:3232
#10 0x00000000004789fd in spdk_bdev_io_complete (bdev_io=0x2000075706c0, status=SPDK_BDEV_IO_STATUS_SUCCESS) at bdev.c:3318
#11 0x0000000000452e7f in bdev_aio_group_poll (arg=0x1f29800) at bdev_aio.c:266
#12 0x00000000004a713f in spdk_thread_poll (thread=0x1f24d10, max_msgs=0) at thread.c:349
#13 0x0000000000465b9c in _spdk_reactor_run (arg=0x1f24dc0) at reactor.c:320
#14 0x0000000000465eb2 in spdk_reactors_start () at reactor.c:421
#15 0x0000000000464483 in spdk_app_start (opts=0x7ffcb01f1c40, start_fn=0x405481 <spdk_startup>, arg1=0x0, arg2=0x0) at app.c:680
#16 0x000000000040561c in main (argc=77, argv=0x7ffcb01f1dc8) at iscsi_tgt.c:112
>>> p *task
Cannot access memory at address 0x20002a6b0400
>>> p task->current_datain_offset
Cannot access memory at address 0x20002a6b0548
>>> p task->scsi
Cannot access memory at address 0x20002a6b0400
>>>
```

Sadly it wont let me inspect the state of task...

Here's the output of `bt full`, its like 1000 lines so I made it a separate gist.

https://gist.github.com/yaahallo/3bb650af2af68d56407ad7f1a8e9aa7f

**Comment by shuhei-matsumoto** [comment]
@yaahallo 

Thank you for your update. That stack trace is unexpected to me. I haven't found why SPDK fell in that state. I'll update here if I find anything.

**Comment by yaahallo** [comment]
Okay, I'm currently in the process of reproducing the issue with a breakpoint in an if statement rather than the assert so hopefully I should be able more information for you rather than a bunch of cannot access memory errosr.

**Comment by yaahallo** [comment]
Just wanted to touch base on this and see where you guys are on this issue and what your thoughts are.

Are you guys still looking into this issue / working on it? Also, do you think that the fact that done tasks are still in the queue at the start of `_spdk_iscsi_conn_abort_queued_datain_task` is a serious issue?  For us we're going to move forward with the assumption that we're fine with the flow control fix already discussed and stop looking into this and approach any instability that we run into as a separate issue.

**Comment by shuhei-matsumoto** [comment]
@yaahallo @jimharris 
I also stopped working on this. Let's reopen if necessary. Thank you for all of your help
